src/core/components/layout/components/columns/styles.ts

```diff
@@ -117,35 +122,99 @@ const generateWidthCSS = (width: number | number[]) => {
 		];
 
 		return width.reduce((styles, w, i) => {
-			return css`
+			return `
 				${styles}
 				${from[breakpoints[i]]} {
 					${calculateWidth(w)};
 				}
 			`;
-		}, css``);
+		}, ``);
 	}
 
 	return calculateWidth(width);
 };
 
-export const column = (width?: number | number[]) => {
-	let flex;
-	let widthCSS;
-
-	if (width == null || (Array.isArray(width) && width.length === 0)) {
-		// If no width is specified, allow the column to grow
-		flex = 1;
-		widthCSS = css``;
-	} else {
-		// If a width is specified, don't allow column to grow. Use the width property
-		flex = '0 0 auto';
-		widthCSS = generateWidthCSS(width);
+const generateSpanCSS = (span: number | number[]) => {
+	if (Array.isArray(span)) {
+		const breakpoints: Breakpoint[] = [
+			'mobile',
+			'tablet',
+			'desktop',
+			'leftCol',
+			'wide',
+		];
+
+		return span.reduce((styles, w, i) => {
+			console.log(`
+			${styles}
+			${from[breakpoints[i]]} {
+					${calculateSpan(w)};
+				}
+			`);
+			return `
+				${styles}
+				${from[breakpoints[i]]} {
+					${calculateSpan(w)};
+				}
+			`;
+		}, ``);
 	}
 
-	return css`
-		box-sizing: border-box;
-		flex: ${flex};
-		${widthCSS};
+	return calculateSpan(span);
+};
+
+const columnBreakpoints: Array<ColumnBreakpoint> = [
+	{ totalColumns: 4, rule: until.tablet },
+	{ totalColumns: 12, rule: between.tablet.and.leftCol },
+	{ totalColumns: 14, rule: between.leftCol.and.wide },
+	{ totalColumns: 16, rule: from.wide },
+];
+
+const calculateSpan = (span: number) => {
+	const columnBreakpointCss = columnBreakpoints.reduce(
+		(acc, cur: ColumnBreakpoint) => {
+			if (span === 0)
+				return `
+					display: none;
+				`;
+
+			const inferredWidth = span / cur.totalColumns;
+			const cappedWidth = inferredWidth < 1 ? inferredWidth : 1;
+			const cssForBreakpoint = calculateWidth(cappedWidth);
+
+			return `
+				${acc};
+				${cur.rule} {
+					${cssForBreakpoint};
+					display: block;
+				}
```

Comment:
I don't think we need the media query here as it's already being generated in [`generateSpan`](https://github.com/guardian/source/pull/760/files#diff-61a31ebdb74d46042d5af57ee18f291987ca2be4bea6ebc25d0ef613c5678e48R156). Does `calculateSpan()` need to receive the `totalColumns`?

Reply:
If we're talking about the same thing (`cur.rule` above)- I think we may still need this one - what this does is reflect the fact that _n_ columns have a different meaning for the width at different breakpoints. E.g. '2' columns means 50% of the screen at mobile size (where there are 4 columns total) but only ~16% at desktop size (where there are 12 total columns). 

Essentially a rule saying that an element should span '2' columns is really generating 4 sets of rules - one for each column-relevant breakpoint ( <tablet, tablet & desktop, leftCol & wide). Without `cur-rule` - '2' columns would be treated the same at every breakpoint.

e.g. removing `cur.rule`:
https://media.giphy.com/media/M2pTPdvNNDopNI4ZBP/giphy.gif

Reply:
I see. Isn't that what [this media query](https://github.com/guardian/source/pull/760/files#diff-61a31ebdb74d46042d5af57ee18f291987ca2be4bea6ebc25d0ef613c5678e48R148) ought to be doing? At the moment it looks like the media query in `generateSpan` is wrapping the one in `calculateSpan`.

Currently if you `console.log()` the output of `generateSpanCSS([0, 0, 2])`, you get CSS that looks like this:

```css
@media (min-width: 320px) {
	@media (max-width: 739px) {
		display: none;
	}
	@media (min-width: 740px) and (max-width: 1139px) {
		display: none;
	}
	@media (min-width: 1140px) and (max-width: 1299px) {
		display: none;
	}
	@media (min-width: 1300px) {
		display: none;
	}
}
```

I guess I'm wondering if this output could be streamlined?

Reply:
Good spot Simon - I may have to do some more thinking about that.

I think the recently removed - 
``` 
if (span === 0)
  return `
    display: none;
  `; 
```

- might have exacerbated this particular case (I'm not totally sure without testing though). But there would probably be unnecessary queries in other cases too, e.g. wide breakpoint rules for a mobile breakpoint specified in an array of breakpoint values, e.g. `[0, 0, 2]`

I think streamlining this will be quite complicated - I'll have a look early next week.

Reply:
Okay, thanks @rhystmills. If you don't have time or it looks like too big a task, we can put a pin in it for now and circle back around. I think making the API available is more important than implementing perf optimisations right now 🙂

Reply:
Re-adding the `return 'display: none'` does help with that particular case (multiple nested `display: none` rules), and reduces the amount of generated css overall - so I have added it back in.

I think there are definitely other cases where smarter logic could help remove redundant rules. 

```css
@media (min-width: 320px) {
display: none;
}

@media (min-width: 740px) {
display: none;
}

@media (min-width: 980px) {
    @media (max-width: 739px) {
      width: calc((100%) * 0.5 - 20px);
      /* Reset value that might have been set at a lower breakpoint */
      display: block;
    }

    @media (min-width: 740px) and (max-width: 1139px) {
      width: calc((100%) * 0.16666666666666666 - 20px);
      /* Reset value that might have been set at a lower breakpoint */
      display: block;
    }

    @media (min-width: 1140px) and (max-width: 1299px) {
      width: calc((100%) * 0.14285714285714285 - 20px);
      /* Reset value that might have been set at a lower breakpoint */
      display: block;
    }

    @media (min-width: 1300px) {
      width: calc((100%) * 0.125 - 20px);
      /* Reset value that might have been set at a lower breakpoint */
      display: block;
    }
}
```
For example above we can see:
- Two consecutive `display: none` breakpoints that could be merged into one
- A `@max-width: 739px` media query that can never have an effect because it's within a `@min-width: 980px` media query.

Also - console logging the css strings shows they have a lot of whitespace in it (as a consequence of multiline css strings in the code), it might be nice to make more readable outputted css by changing how we do this in the future.

Reply:
Perhaps those performance improvements would be good for a future PR?

Reply:
Thanks for investigating, solid work 💪 

I agree we should park these optimisations for now and decouple them from this PR. We're already covering a lot of ground.